core/src/main/resources/hudson/PluginManager/_table.js

```diff
@@ -10,46 +10,214 @@ function checkPluginsWithoutWarnings() {
 
 Behaviour.specify("#filter-box", '_table', 0, function(e) {
     function applyFilter() {
-        var filter = e.value.toLowerCase().trim();
-        var filterParts = filter.split(/ +/).filter (function(word) { return word.length > 0; });
-        var items = document.getElementsBySelector("TR.plugin");
-        var anyVisible = false;
-        for (var i=0; i<items.length; i++) {
-            if ((filterParts.length < 1 || filter.length < 2) && items[i].hasClassName("hidden-by-default")) {
-                items[i].addClassName("hidden");
-                continue;
-            }
-            var makeVisible = true;
-
-            var content = items[i].innerHTML.toLowerCase();
-            for (var j = 0; j < filterParts.length; j++) {
-                var part = filterParts[j];
-                if (content.indexOf(part) < 0) {
-                    makeVisible = false;
-                    break;
-                }
-            }
-            if (makeVisible) {
-                items[i].removeClassName("hidden");
-                anyVisible = true;
+        clearTimeout(debounce);
+        // debounce reduces number of server side calls while typing
+        debounce = setTimeout(function() {
+            var pluginsTable = document.getElementById('plugins');
+
+            // only on available tab
+            if (pluginsTable.dataset.page === 'available') {
+                view.availablePlugins(e.value.toLowerCase().trim(), 20, function (plugins) {
+                    var availablePlugins = JSON.parse(plugins.responseObject());
+
+                    var tbody = pluginsTable.querySelector('tbody');
+
+                    function clearOldResults() {
+                        if (pluginsTable.dataset.hasadmin !== 'true') {
+                            tbody.innerHTML = '';
+                        } else {
+                            var rows = tbody.querySelectorAll('tr');
+                            if (rows) {
+                                rows.forEach(function (row) {
+                                    var input = row.firstChild.firstChild;
+                                    if (input.checked !== true) {
+                                        row.remove();
+                                    }
+                                })
+                            }
+                        }
+                    }
+
+                    clearOldResults();
+
+                    availablePlugins.forEach(function (plugin) {
+                        var htmlTableRowElement = document.createElement('tr');
+                        htmlTableRowElement.classList.add('plugin');
+
+                        function createCheckBox() {
```

Comment:
shout if any one knows a more Jenkins' way to do this.

It's not very nice, but huge performance gain